src/encord_active/server/routers/project.py

```diff
@@ -394,28 +467,72 @@ def _get_or_create_tag_hash(name: str) -> uuid.UUID:
                         continue
                     new_label_tag_uuids.add(tag_hash)
                     label_exists.add(dup_key2)
-                    if tag_hash in existing_label_tags:
+                    if tag_hash in existing_label_tags.get((du_hash, frame, annotation_hash), set()):
                         continue
-                    sess.add(
-                        ProjectTaggedAnnotation(
-                            project_hash=project_hash,
-                            du_hash=du_hash,
+                    annotation_tags_to_add.append(
+                        AnnotationTagAdd(
+                            project_hash=project_hash.hex,
+                            du_hash=du_hash.hex,
                             frame=frame,
+                            tag_hash=tag_hash.hex,
                             object_hash=annotation_hash,
-                            tag_hash=tag_hash,
                         )
                     )
-                for tag_hash in existing_label_tags.difference(new_label_tag_uuids):
-                    tag_to_remove = sess.exec(
-                        select(ProjectTaggedAnnotation).where(
-                            ProjectTaggedAnnotation.project_hash == project_hash,
-                            ProjectTaggedAnnotation.du_hash == du_hash,
-                            ProjectTaggedAnnotation.frame == frame,
-                            ProjectTaggedAnnotation.object_hash == annotation_hash,
-                            ProjectTaggedAnnotation.tag_hash == tag_hash,
-                        )
-                    ).first()
-                    sess.delete(tag_to_remove)
+
+                annotation_tags_to_remove.update(
+                    set(
+                        [
+                            (du_hash, frame, annotation_hash, tag_hash)
+                            for tag_hash in existing_label_tags.get(
+                                (du_hash, frame, annotation_hash), set()
+                            ).difference(new_label_tag_uuids)
+                        ]
+                    )
+                )
+
+        # Delete left over data and annotation tags
+        if data_tags_to_add:
+            sess.execute(
+                text(
```

Comment:
change this to sess,add(ProjectTaggedX()) - sqlalchemy will automatically batch the changes. May be faster.